### ISSUE TITLE
[Proposal] Union types with type aliases

### ISSUE THREAD

**Issue by Artur-A** [Microsoft/TypeScript#11065]
**TypeScript Version:**  2.0 rc
Great work has been done for union types and type narrowing. The following example with created type from string literals works great:
**Code**

``` ts
interface ActionA {
   Key: "ActionAOccured";
   Data: number;
}
interface ActionB {
   Key: "ActionBOccured";
   // some B data
}

ProcessAction(action: ActionA | ActionB){
   switch(action.Key){
      case "ActionAOccured":
          // here we have support of ActionA objet
          // can work with action.Data and have static analysis
  }
}
```

Using strings for such comparison in code often leads to typos and unexpected behaviour.
I tried to use an "ideal" variant with `type alias`, but it does not work:

``` ts
namespace Keys {
    export type ActionAOccured = "ActionAOccured";
    export type ActionBOccured = "ActionBOccured";
}

ProcessAction(action: ActionA | ActionB){
   switch(action.Key){
      case Keys.ActionAOccured:<----
         Error: Property 'ActionAOccured' does not exist on type 'typeof Keys'**
  }
}
```

Also even use of a usual string in switch does not produce an error if there is a wrong "string literal type":

``` ts
ProcessAction(action: ActionA | ActionB){
   switch(action.Key){
      case "NonexistingActionKey"
         // Compiles without a problem, no check, although tsc knows all about Key type
  }
}
```

This feature is very popular, especially in React with dispatcher and action switching. Now it works only with string literal types (potential misspelling) and with `instanceof`/casting, which is even less inconvenient to write and read.

**Comment by mhegazy** [comment]
You can emulate what you want today using this pattern:

```ts
namespace Keys {
    export type ActionAOccured = "ActionAOccured";
    export const ActionAOccured: ActionAOccured = "ActionAOccured";
    export type ActionBOccured = "ActionBOccured";
    export const ActionBOccured: ActionBOccured  = "ActionBOccured";
}
```

A string enum should just desugars to that.

**Comment by mhegazy** [comment]
string enums are tracked by https://github.com/Microsoft/TypeScript/issues/3192 (and https://github.com/Microsoft/TypeScript/issues/1206).

**Comment by Artur-A** [comment]
Nice, thank you @mhegazy,
string enum will be a great enhancement.
